### ISSUE TITLE
Estimating target bitrate with RetmoteBitrateEstimator doesn't work well

### ISSUE THREAD

**Issue by jaykim305** [aiortc/aiortc#170]
I am doing  webrtc experiments in a network emulated environment where bandwidth is congested and fluctuated, thus my application needs the rate controlling feature to adapt to the bandwidth.    

**However, the target bitrate estimate from RemoteBitrateEstimator value doesn't change that much compared to the bandwidth.**  

My application consists of webrtc based [client---server] using ```aiortc/examples/videostream-clie/cli.py and aiortc/examples/server/server.py ``` 

As I understand, rate controlling algorithm in aiortc is a **GCC algorithm (goog-REMB)** and it generally works as follows, 

1. When RTCRTPReceiver(server in my app) receives video, it calculates target bitrate for sender by using RemoteBitrateEstimator and sends the target bitrate by REMB packet.

2. Then RTCRTPSender(client in my app) receives the REMB packet and sets the target bitrate to the encoder.

I noticed that the code structure in `rate.py` is very similar to **`aimd_rate_control.cc`**
in webrtc native code.   
However, still it was not same as original code; there was no loss based controller implemented in sender side and seems to be missing several implementation details.

My questions are, 

- **Is airotc rate controlling algorithm fully implemented enough compared to webrtc ?**
- **Have you tested compared the rate controlling performance with native webrtc ??**

I want some guarantee of aiortc's reasonable rate controlling algorithm for my further implementations

This is my simulation result graph for reference and it seems that target bitrate estimate value hardly changes.

![bw_result](https://user-images.githubusercontent.com/20180906/56789537-1409ef80-683d-11e9-9ae4-b686a2381748.JPG)

**Comment by jlaine** [comment]
Could you provide some more information about what issue you are actually encountering, as the graphs don't tell me much? To the best of my knowledge the estimation part is correctly implemented. As you pointed out it is very similar to the webrtc codebase.

Could we also dial the tone down a notch, "you guys" is a single person as I don't receive many meaningful code contributions.

**Comment by jaykim305** [comment]
I am really sorry for my tone. I wasn't sure to whom I should approach the question. 

I am not sure if the  rate controlling is working properly, because the target bitrate doesn't really change while the bandwidth is fluctuating.  In fact, it converges to constant value as shown in the graph.

Also, to my knowledge, considering loss in estimating bitrate is major part in google GCC algorithm but I couldn't see the implementations in the code. 

I was wondering have you tested the rate controlling algorithm and confirmed that it works just fine and similarly as webrtc does. 

By the way, I really appreciate you work! Thank you for your  contribution and quick reply.

**Comment by jlaine** [comment]
I don't currently have much time for a deep dive into the code so here is what I recall from implementing the rate estimation code:

- the goal is not to track the *used* bandwidth but the *available* bandwidth. This is an important distinction: look at the `RemoteBitrateEstimator` class and you'll see there is an `incoming_bitrate` which *does* track the effectively used bandwidth, but this is *not* what is directly used as the estimated bandwidth.

- the estimated available bandwidth is primarily driven by detecting "bandwidth overuse". Bandwidth overuse is detected by observing changes in the delta between a packet's sent and arrival timestamps. Roughly speaking the underlying assumption is that if you are exceeding the available bandwidth, buffers will start filling up somewhere on the network and packets will take longer and longer to arrive.

- loss detection is *not* used at all, this is purely a delay-based congestion control. This would be an interesting addition, if you are are familiar with the webrtc.org could you point me to where this is used?

**Comment by quito418** [comment]
Hi jlaine

I think Jaykim305 is asking about this part in webrtc source.

**URL:** https://chromium.googlesource.com/external/webrtc/+/branch-heads/63/modules/bitrate_controller/send_side_bandwidth_estimation.cc#277

**Function Name:** void SendSideBandwidthEstimation::UpdateEstimate(int64_t now_ms) 

I think this part is implementing the loss-based controller in Figure 1 of GCC paper ( https://c3lab.poliba.it/images/6/65/Gcc-analysis.pdf )

Thank you for your hard work.

With regards

**Comment by jaykim305** [comment]
Just for your information, 
previously, I had one issue before this problem and solved it. 
The issue was that rtcrtpsender was not getting REMB feedback packet from rtcrtpreceiver eventhough rtcrtpreceiver was keep sending it.

This was because the receiver sets the `media_ssrc=0` of the packet like below
                      
https://github.com/aiortc/aiortc/blob/0310cd8486b38d8d8f295650253224247c43ed42/aiortc/rtcrtpreceiver.py#L382
, but the ssrc value actually is not 0 at all. (The ssrc value for each peer is set when exchanging SDP)

When I changed the `media_ssrc=0` part to `media_ssrc=packet.ssrc`,
I was able to receive REMB feedback packet from rtcrtpreceiver.

**Comment by jlaine** [comment]
> When I changed the `media_ssrc=0` part to `media_ssrc=packet.ssrc`,
> I was able to receive REMB feedback packet from rtcrtpreceiver.

OK that looks like an interesting find.

I just checked what Firefox and Chrome send - and can confirm they too set `media_ssrc` to 0, so I don't think your fix is correct. There does however be a problem with the way these RTCP are routed back to the RTCRtpSender:

https://github.com/aiortc/aiortc/blob/0310cd8486b38d8d8f295650253224247c43ed42/aiortc/rtcdtlstransport.py#L271

Here we can see the routing is done on `media_ssrc` which we know will be 0. I need to check the specs to see how we are supposed to correlate such packets.

**Comment by jlaine** [comment]
@jaykim305 routing the REMB packets to RTCRtpSender is now fixed in 48a950f4629c370b16f2abb84d205e42abce3f61

**Comment by HoustonHuff** [comment]
Hello, I saw this old issue and was hoping to re-open it in the context of a static, non-GCC target bitrate.

Like the original poster I've been working on WebRTC experiments using a namespace network and running the browser receiver through Chrome, and I similarly I have been getting a live frame stream bitrate that fluctuates and doesn't seem bound by the target bitrate. I've verified that my version seems to have the changes implemented in 48a950f. I'll admit that my understanding of media_ssrc is limited, so maybe there's another aspect that I've not addressed in my setup that I need to in order to see the stable bitrate that I was expecting. I've attempted both with and without the GCC estimator, though from my understanding with a similar result, though as far as I'm aware that shouldn't be the determining factor in any case.

Any guidance or help in troubleshooting would be very appreciated.
